---
title: "expr: Longitudinal Analysis"
author: "Xiqi Li"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  rmdformats::readthedown:
    toc_depth: 4
    tabset: true
    number_sections: false
    self_contained: true
    thumbnails: false
    lightbox: false
params:
  DGE_logFC_thresh: 
    label: "logFC threshold for significance:"
    input: slider
    min: 0
    max:  2
    step: 0.05
    value: !r log2(1.5)
  Nit:
    label: "Set number of iterations for consensus clustering on expression:"
    value: 50
    input: slider
    min: 20
    max: 2000
    step: 10
  NofK:
    label: "Select number of clusters K for clustering on expression:\n\t If Left blank optimum K will be calculated."
    value: 2
    input: numeric
    min: 2
    max: !r round(sqrt(dim(paired_pt_info)[1]))
    step: 1
  tracks:
    label: "Include as heatmap tracks:"
    value: !r tracks.d
    choices: !r colnames(paired_pt_info)
    input: select
    multiple: TRUE
  OS_thresh: 
    label: !r sprintf("Threshold to split long/short overall survival (min=%s,max=%s,median=%s):\n\tSkip if data unavailable",OSthresh[1],OSthresh[3],OSthresh[2])
    input: numeric
    value: !r NA
    min: !r as.numeric(OSthresh[1])
    max: !r as.numeric(OSthresh[3])
    step: 0.01
  PFS_thresh: 
    label: !r sprintf("Threshold to split long/short progression-free survival (min=%s,max=%s,median=%s):\n\tSkip if data unavailable",PFSthresh[1],PFSthresh[3],PFSthresh[2])
    input: numeric
    value: !r NA
    min: !r as.numeric(PFSthresh[1])
    max: !r as.numeric(PFSthresh[3])
    step: 0.01
  response_thresh: 
    label: "Define binary reponse categories: \n\tSkip if data unavailable or using quantitative response"
    input: select
    choices: ["stable disease as good response","stable disease as bad response",NA]
    value: NA
  showCI: 
    label: "Show confidence interval in Kaplan Meier plot."
    value: FALSE
editor_options: 
  chunk_output_type: console
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  # comment = "#>",
  tidy=TRUE,
  tidy.opts=list(arrow=TRUE, indent=2,width.cutoff=60),
  message = FALSE,
  warning = FALSE,
  echo = FALSE
)

require(expr)
require(dplyr)
require(htmltools)
require(DT)
require(plotly)

load("clean_batchexamined_logRNA.RData")
if (correctionMetod=="limma"){
  log2_expressions=limma_rbe_log2_protein_expressions
  cat("<br>Using limma batch-corrected expression matrix...<br>")
}else if(correctionMetod=="combat"){
  log2_expressions=combat_rbe_log2_protein_expressions
  cat("<br>Using ComBat batch-corrected expression matrix...<br>")
}else{
  log2_expressions=log2_clean_protein_expressions
  cat("<br>Using uncorrected expression matrix...<br>")
}
rm(list = setdiff(ls(),c("params","clean_RNA_sample_info","log2_expressions","metaHeaders")))

tpheader="time_point"
clean_RNA_sample_info[,tpheader]=gsub("[-_ \t\n\r\v\f]","",clean_RNA_sample_info[,tpheader])

paired_RNA_sample_info=clean_RNA_sample_info[clean_RNA_sample_info$MRNsurr%in%names(which(table(clean_RNA_sample_info$MRNsurr)>1)),] %>% arrange(time_point)%>%arrange(order(gtools::mixedorder(MRNsurr)))
DGE=DGE.DT=DGE.plots=volcanoPlots=gseaRess=gseaRes.sig=split_groups=gseaPlots=list()
```

## Data set summarized by time point
```{r dataset-summary,out.width="100%"}
paired_pt_info=reshape2::dcast(paired_RNA_sample_info[,c("MRNsurr","time_point","Sample_ID")],MRNsurr~time_point,value.var="Sample_ID",fun.aggregate = function(x) paste(unique(x),collapse = ";")) %>% tibble::column_to_rownames("MRNsurr")
num.ind=colnames(paired_RNA_sample_info)[sapply(paired_RNA_sample_info, is.numeric)]
for (header in colnames(paired_RNA_sample_info)[!grepl("Sample_ID|MRN|full|name",colnames(paired_RNA_sample_info),ignore.case = T)]){
  tmp=reshape2::dcast(paired_RNA_sample_info[,c("MRNsurr","time_point",header)],MRNsurr~time_point,value.var=header,fun.aggregate = function(x) paste(unique(x),collapse = ";")) %>% tibble::column_to_rownames("MRNsurr")
  if(all(apply(tmp, 1, function(x) length(unique(na.omit(x[x!=""]))))==1)){
    paired_pt_info[[header]]=apply(tmp,1,function(x) unique(na.omit(x[x!=""])))
  }
}
num.ind=num.ind[num.ind%in%colnames(paired_pt_info)]
paired_pt_info[,num.ind]=apply(paired_pt_info[,num.ind],2,as.numeric)
paired_pt_info[paired_pt_info==""]=NA
paired_pt_info=paired_pt_info[gtools::mixedorder(rownames(paired_pt_info)),]

tags$h5("Patients with paired samples",style = "color:black;")
paired_pt_info %>% datatable(
  filter = "top",
  options = list(
    autoWidth = TRUE, columnDefs = list(list(targets = "_all")),
    # scrollX= T,scrollY="650px",scrollCollapse=T,
    # scrollCollapse=T,
    initComplete = JS("function(settings, json) {","$(this.api().table().header()).css({'font-size': '12px'});","}"))
  ) %>%
  DT::formatRound(columns = names(which(sapply(.$x$data,is.numeric))), digits = 2)%>%
  formatStyle(columns = colnames(.$x$data), `font-size` = '12px')
# tags$div(class="scroll",tmp,style='width:100%;height=350px')

```

## Comparison between Baseline group and TP2 group
```{r DGE-baseline-VS-TP2,out.width="100%"}
library(ggrepel)
require(fgsea)
load(system.file("extdata/protein_coding_ensemble2symbol.RData",package = "expr"))
load(system.file("extdata/pathway.nano_hallmarkv2022.1.RData",package = "expr"))


#read oncokb
oncoKB=read.csv(system.file("extdata/cancerGeneList_oncoKB_lastupdate05192023.tsv",package="expr"),sep = "\t",header = T,check.names = F)
rownames(oncoKB)=oncoKB$`Hugo Symbol`
oncoKB=oncoKB%>%mutate(
    `Is Oncogene`=replace(`Is Oncogene`,`Is Oncogene`=="Yes",T),
    `Is Oncogene`=replace(`Is Oncogene`,`Is Oncogene`=="No",F),
    `Is Oncogene`=as.logical(`Is Oncogene`),
    `Is Tumor Suppressor Gene`=replace(`Is Tumor Suppressor Gene`,`Is Tumor Suppressor Gene`=="Yes",T),
    `Is Tumor Suppressor Gene`=replace(`Is Tumor Suppressor Gene`,`Is Tumor Suppressor Gene`=="No",F),
    `Is Tumor Suppressor Gene`=as.logical(`Is Tumor Suppressor Gene`),
    `OncoKB Annotated`=replace(`OncoKB Annotated`,`OncoKB Annotated`=="Yes",T),
    `OncoKB Annotated`=replace(`OncoKB Annotated`,`OncoKB Annotated`=="No",F),
    `OncoKB Annotated`=as.logical(`OncoKB Annotated`)
    )
#read cancer gene census
cancerGeneCensus=read.csv(system.file("extdata/concensus_Census_allWed_Jan_18_20_33_56_2023.tsv",package="expr"),sep = "\t",header = T,check.names = F)
rownames(cancerGeneCensus)=cancerGeneCensus$`Gene Symbol`
cancerGeneCensus$Hallmark=cancerGeneCensus$Hallmark %>% replace(.,.=="Yes",T) %>% replace(.,.=="",F) %>% as.logical()

#load pathways
phenotypes=c("group_comparison","paired_comparison")
for(phenotype in phenotypes){

if(phenotype=="group_comparison"){
  group=as.factor(clean_RNA_sample_info[,"time_point"])
  contrast=sprintf("%s-%s",sort(levels(group))[2],sort(levels(group))[1])
  group=data.frame(group)
  dimnames(group)=list(rownames(clean_RNA_sample_info),phenotype)
  group=data.frame(Sample_ID=clean_RNA_sample_info$Sample_ID,group)
  group=group[complete.cases(group),]

  sink(NULL)
    comp_limma=dge_limma(
    expressions = log2_expressions[,group$Sample_ID], is_rawcount = F,is_logged = T,normalize = F,
    sample_frequency_threshold=0.5,
    clinic_info = group,
    ID_col = "Sample_ID",
    group_col = phenotype,
    contrasts = c(contrast),
    method ="limma_trend")
  sink()
}else if(phenotype=="paired_comparison"){
  group=as.factor(paired_RNA_sample_info[,"time_point"])
  contrast=sprintf("%s-%s",sort(levels(group))[2],sort(levels(group))[1])
  group=data.frame(group)
  dimnames(group)=list(rownames(paired_RNA_sample_info),phenotype)
  group=data.frame(paired_RNA_sample_info[,c("Sample_ID","MRNsurr")],group)
  group=group[complete.cases(group),]
  sink(NULL)
    comp_limma=dge_limma(
    expressions = log2_expressions[,group$Sample_ID], is_rawcount = F,is_logged = T,normalize = F,
    sample_frequency_threshold=0.5,
    clinic_info = group,
    ID_col = "Sample_ID",
    block_col = "MRNsurr",
    group_col = phenotype,
    contrasts = c(contrast),
    method ="limma_trend")
  sink()
}
split_groups[[phenotype]]=setNames(group[[phenotype]],group[["Sample_ID"]])

# format stats table
stats=comp_limma$statistics %>% arrange(get(sprintf("%s:P.Value",contrast)))
stats=cbind(stats,protein_coding_ensemble2symbol[match(rownames(stats),protein_coding_ensemble2symbol$Gene_Symbol),-1])

tmp=apply(oncoKB[stats$Gene_Symbol,c("OncoKB Annotated","Is Oncogene","Is Tumor Suppressor Gene")], 1, function(x) {
  x[1]=!x[1]
  paste(c("not annotated","oncogene","TSG")[x],collapse = "; ")})
tmp=tmp %>% replace(.,.=="NA; NA; NA",NA) %>% replace(.,.=="","other")
stats=cbind(stats,oncoKB=tmp)
tmp=apply(cancerGeneCensus[stats$Gene_Symbol,c("Tier","Hallmark")], 1, function(x)
  paste(paste("Tier",x[1],sep=""),"Hallmark"[x[2]]))
tmp=tmp %>% replace(.,.=="TierNA NA",NA) %>% trimws()
stats=cbind(stats,cancerGeneCensus=tmp)

DGE[[phenotype]]=stats
sigRow=stats[[sprintf("%s:P.Value",contrast)]]<0.05 & abs(stats[[sprintf("%s:logFC",contrast)]])>params$DGE_logFC_thresh
DGE.DT[[phenotype]]= stats[sigRow,] %>%
  datatable(
    filter = "top",
    options = list(
    # scrollX=T,scrollY="650px",scrollCollapse=T,
    initComplete = JS("function(settings, json) {","$(this.api().table().header()).css({'font-size': '12px'});","}"),
    autoWidth = TRUE,
    columnDefs = list(list(width = '200px', targets = which(colnames(stats)=="Gene_Info"))))
    ) %>%
  DT::formatRound(columns = names(which(sapply(.$x$data,is.numeric))), digits = 2)%>%
  formatStyle(columns = colnames(.$x$data), `font-size` = '12px')

  # box plots
  if(sum(sigRow)>0){
    DGE.plots[[phenotype]]=as_widget(ggplotly(
      MetabClean::plotBoxPlot(
        data_mx = log2_expressions,sampleAttr = group,
        MOI = (stats[sigRow,]%>% rownames())[1:min(12,sum(sigRow))],
        header4x = phenotype,header4ID = "Sample_ID",header4color = phenotype) +
        theme(legend.position = "none") +
        labs(title = "Top differentially expressed genes")
      ))
  }

  # generate volcano plots
  dff=stats %>%
    mutate(diffexpressed=ifelse(
      get(sprintf("%s:P.Value",contrast))<0.05,
      ifelse(
        get(sprintf("%s:logFC",contrast))>params$DGE_logFC_thresh,"Up",
        ifelse(
          get(sprintf("%s:logFC",contrast))<(-params$DGE_logFC_thresh),
          "Down",
          "No")),
      "No")) %>%
    mutate(`-log10(P)`=-log10(get(sprintf("%s:P.Value",contrast)))) %>%
    mutate(label=Gene_Symbol)
  dff$label[!(dff$`-log10(P)`>(-log10(0.05))&dff$diffexpressed!="No")]=NA

  volcanoPlots[[phenotype]]=ggplot(
    data=dff,
    aes(x=.data[[sprintf("%s:logFC",contrast)]],
        y=`-log10(P)`,
        col=diffexpressed, label=label)) +
    geom_point() +
    theme_minimal() +
    geom_text_repel() +
    scale_color_manual(name="expression",values = c("dodgerblue3", "gray50", "firebrick3")) +
    geom_vline(xintercept=c(-params$DGE_logFC_thresh, params$DGE_logFC_thresh), linetype=2) +
    geom_hline(yintercept=-log10(0.05), linetype=2) +
    labs(title = sprintf("Differentially expressed genes (p<0.05, |log2FC|> %s)",params$DGE_logFC_thresh))+
    xlab(sprintf("%s:logFC",contrast))

  # GSEA
    # generate gsea table
  rank=stats %>% #filter(p_value<0.05,abs(Difference_of_means)>paramss$meanDiff.thresh) %>%
    arrange(get(sprintf("%s:logFC",contrast))) %>% select(sprintf("%s:logFC",contrast)) %>%
    tibble::rownames_to_column("x") %>% tibble::deframe()
  gseaRess[[phenotype]]=gseaRes=fgsea(
    pathways=pathway,
    stats=rank,
    minSize = min(sapply(pathway, length)),
    gseaParam = 1)
  gseaRes.sig[[phenotype]]=gseaRes %>% filter(pval<0.05) %>% arrange(pval) %>%
    datatable(
      filter = "top",
      options = list(
        # scrollX="100%",scrollY="650px",scrollCollapse=T,
        initComplete = JS("function(settings, json) {","$(this.api().table().header()).css({'font-size': '12px'});","}")),
    rownames = FALSE,
    # caption = htmltools::tags$caption(
    #   style = 'caption-side: top; text-align: left; color:black; font-size:150% ;',
    #   sprintf("GSEA on %s: %s", phenotype, contrast))
    ) %>%
    DT::formatRound(columns = names(which(sapply(.$x$data,is.numeric))), digits = 2)%>%
    formatStyle(columns = colnames(.$x$data), `font-size` = '12px')

  # generate gsea plots
  fgseaRes=gseaRes %>% filter(pval<0.05) %>% arrange(NES)
  gseaPlots[[phenotype]]=plotGseaTable(pathway[fgseaRes$pathway], rank, fgseaRes,
              gseaParam=1)
}

img.gsea=lapply(gseaPlots, function(x) {
  htmltools::plotTag(
  {x},
  sprintf("%s GSEA: significant pathway", names(x)), width = 700, height = 600, pixelratio=3)
})

img.DGE=lapply(volcanoPlots, function(x) {
  htmltools::plotTag(
  {x},
  sprintf("%s DGE: volcano plot", names(x)), width = 600, height = 400, pixelratio=3)
})

tabs <- lapply(phenotypes, function(name) {
  shiny::tabPanel(name, htmltools::tagList(
    list(
    img.DGE[[name]],
    tags$h5(sprintf("DGE on %s: %s",phenotype,contrast),style="color:black;"),
    DGE.DT[[name]],
    DGE.plots[[name]],
    tags$hr(style = "border-top: 1px solid #ccc;"),
    tags$h5(sprintf("GSEA on %s: %s", phenotype, contrast),style="color:black;"),
    img.gsea[[name]],
    gseaRes.sig[[name]]
    )))
})
save(gseaRess,pathway,DGE,split_groups,file = sprintf("gsea_%s.RData","TPgroup"))
tagList(do.call(shiny::tabsetPanel, tabs))
```

## Analysis on delta profiles
If a patient has more than one sample for the same time point, the first sample is used for calculating delta.
```{r delta-data-load}
rm(list = setdiff(ls(),grep("log2_expressions|params|paired_pt|clean",ls(),value=T)))
paired_pt_info=paired_pt_info[complete.cases(paired_pt_info[,c(1,2)]),]
delta_log2_expressions=apply(paired_pt_info[,c(1,2)],1,function(x) {
  log2_expressions[,strsplit(x[2],";")[[1]][1]]-log2_expressions[,strsplit(x[1],";")[[1]][1]]
  })
rownames(delta_log2_expressions)=rownames(log2_expressions)
log2_expressions.0=log2_expressions
log2_expressions=delta_log2_expressions
clean_RNA_sample_info.0=clean_RNA_sample_info
clean_RNA_sample_info=paired_pt_info
clean_RNA_sample_info$Sample_ID=rownames(clean_RNA_sample_info)
save(delta_log2_expressions,paired_pt_info,file = sprintf("delta_log2.RData"))
```

### Unsupervised clustering
An in-house method is used to achieve **stable consensus clustering** in 3 steps: 

1. Determine best number of clusters between 2 and rounded square root of sample numbers;
2. Evaluate partitions generated by 13 clustering methods and exclude the most unlikely methods (sum of rand similarity to all other methods <0) from step 3.
3. Calculate pair-wise sample clustering probability from bootstrap iterations of partitioning. In each iteration, all methods in step 2 are used to cluster a randomly selected subset of samples.
  
Correlation plot shows similarity of partitions by all methods in step 2.
The heat map dendrogram across samples resembles distance matrix of pair-wise sample clustering probability.
```{r clustering, include=FALSE}
unsupervised_expressions<-prepare_unsupervised_data(log2_expressions,method="MAD",mad_top_n =1000)
  if(is.na(params$NofK)|params$NofK=="NA"){
    getOptK=function(){
    pdf(file = NULL)
    optK=estimate_bestNumberofClusters(
    unsupervised_expressions,
    method="average",max.nc=round(sqrt(ncol(unsupervised_expressions))))$Best.NumberofCluster
    dev.off()
    return(optK)
  }
  optK=suppressWarnings(suppressMessages(getOptK()))
  }else{
  optK=params$NofK
}


getMultiCluster=function(){
  result=multiCluster(
    data = unsupervised_expressions,
    scale="row",
    bestnumberofclusters=optK,
    cluster_methods=c("kmeans","pam","hclust","fuzzy","mclust","apclust","hdbscan","MCL","specc","kkmeans","skmeans","nmf","som")
  )
  return(result)}
result=suppressWarnings(suppressMessages(getMultiCluster()))

cutFUNs=names(result$rand.sim)[sapply(result$rand.sim, function(x) sum(x,na.rm = T)>0)]
InputParams=data.frame(
  `Number of clusters`=optK,
  `Number of iterations`=params$Nit,
  `Faction of features (genes) used per iteration`=0.8,
  `Faction of samples being clustered per iteration`=1,
  check.names = F
)
ccC=consensusCluster(
  data = unsupervised_expressions,
  method="bootstrap",
  subFeatureSize=InputParams$`Faction of features (genes) used per iteration`,
  subSampleSize=InputParams$`Faction of samples being clustered per iteration`,
  cutFUN=cutFUNs,
  nTimes=InputParams$`Number of iterations`,clusters=optK,verbose=F)
a=hclust(as.dist(1-ccC))
ccC=cutree(a,optK)
```

```{r clustering-plots,out.width="100%",fig.height=7,fig.width=10,results="asis"}
require(ComplexHeatmap)
t(InputParams) %>% knitr::kable(align = "ll",caption = "consensus clustering settings") %>% kableExtra::kable_styling("striped")

plotTag(result$rand.plot, "rand similarity", width = 500, height = 500, pixelratio=3)

tags$hr(style = "border-top: 1px solid #ccc;")

# plot heatmap
tmp=as.factor(paste("C",ccC,sep = ""))
heatmap_mRNA<-draw(Heatmap(
    t(scale(t(unsupervised_expressions))),
    name="delta\nz_score",
    top_annotation = getHeatMapAnnotation(
      sampleAttr = clean_RNA_sample_info,data_mx = log2_expressions,
      # tracks=c("OS","vital_status","PFS","event_for_progression","best_response_perc","time_point","gender","best_response")),
      tracks=params$tracks[params$tracks%in%colnames(clean_RNA_sample_info)]),
    show_row_names = F,column_dend_reorder = T,show_column_names = F,
    cluster_columns = a,
    column_split = optK,
    column_title = levels(tmp[a$order]),
    column_names_gp = gpar(fontsize=1)))
# rm(list = setdiff(ls(),c("params","clean_RNA_sample_info","log2_expressions","metaHeaders","ccC","optK")))
rm(list = setdiff(ls(),grep("log2_expressions|params|clean_RNA_sample_info|ccC|optK",ls(),value=T)))

```


#### OS and/or PFS and/or Response on subclasses from clustering results
```{r survival-on-unsupervised-clustering-results,fig.show='show',out.width="100%"}
metaHeader=R.utils::loadToEnv("exprSetup.RData")[["metaHeader"]]
require("survival")
library("survminer")
require(ggpubr)
clean_RNA_sample_info$Cluster=as.factor(paste("C",ccC[clean_RNA_sample_info$Sample_ID],sep = ""))
clean_RNA_sample_info$Cluster=as.factor(clean_RNA_sample_info$Cluster)
plots=list()

metaHeaders=R.utils::loadToEnv("exprSetUp.RData")[["metaHeaders"]]
censoring=setNames(c("vital_status","event_for_progression"),c("OS","PFS"))
censoring=censoring[metaHeaders[names(censoring)]!="NA" & metaHeaders[censoring]!="NA"]

for (duration in names(censoring)){
  censor=censoring[duration]
  call=as.formula(sprintf("Surv(%s, %s) ~ Cluster",duration,censor))
  fit=surv_fit(call,data = clean_RNA_sample_info)
  logRankTest.pw=pairwise_survdiff(call,data = clean_RNA_sample_info,p.adjust.method="none")$p.value
  logRankTest.oa=survdiff(call,data = clean_RNA_sample_info)$pvalue
  HR=list()
  refU=unique(combn(levels(clean_RNA_sample_info$Cluster),2)[1,])
  for (ref in refU){
    dataSurv=clean_RNA_sample_info[,c(duration,censor,"Cluster")]
    dataSurv$Cluster=factor(dataSurv$Cluster,levels = c(ref,levels(dataSurv$Cluster)[levels(dataSurv$Cluster)!=ref]))
    tmp=summary(coxph(call, data =dataSurv))
    tmp=cbind(
      HR=tmp$coefficients[,"exp(coef)"] %>% format(digits=2),
      CI=paste(tmp$conf.int[,"lower .95"]%>% format(digits=2),tmp$conf.int[,"upper .95"]%>% format(digits=2),sep = "-"),
      `p.coxph`=tmp$coefficients[,"Pr(>|z|)"] %>% format(digits=2))
    rownames(tmp)=apply(combn(levels(dataSurv$Cluster),2)[,1:(nlevels(dataSurv$Cluster)-1),drop=F], 2, function(x) paste(x[2],"vs", x[1],sep = ""))
    HR[[ref]]=tmp
  }
  if(nlevels(dataSurv$Cluster)>2){
    HR=Reduce(rbind,HR)
    HR=HR[!duplicated(lapply(strsplit(rownames(HR),"_vs._"), sort)),drop=F]
  }else{HR=HR[[1]]}
  stats=cbind(HR,`p.logrank`=na.omit(as.vector(logRankTest.pw)) %>% format(digits=2))
  stats=rbind(stats,overall=c(NA,NA,NA,logRankTest.oa%>% format(digits=2)))
  # stats.str=readr::format_delim(stats %>% as.data.frame%>% tibble::rownames_to_column("pair"),delim = "\t")
  stats.str=capture.output(stats %>% as.data.frame%>%knitr::kable("simple",align = 'c'))
  stats.str=paste(stats.str[stats.str!=""&!grepl("-----",stats.str)],collapse = "\n")
  ggsurv=ggsurvplot(fit, data = clean_RNA_sample_info,
           palette =cols4all::c4a(palette = "Dark24",n=optK),
           conf.int = params$showCI,
           risk.table = TRUE,
           risk.table.col = "strata",
           risk.table.height = 0.25,
           ggtheme = theme_bw())
  ggsurv$plot=ggsurv$plot +
    labs(title = sprintf("Kaplan-Meier Curve on %s",duration),)+
    annotate("text",x=range(clean_RNA_sample_info[[duration]])[2]*2.9/4,y=0.81,
             label=stats.str,alpha = .7)
    # annotation_custom(tableGrob(stats,theme=ttheme_minimal(base_size = 6)),
    #                   xmin=range(clean_RNA_sample_info[[duration]])[2]*3.8/5,
    #                   xmax=range(clean_RNA_sample_info[[duration]])[2]*4.8/5,
    #                   ymin=0.75,ymax=0.90)
  # print(ggsurv)
  plots[[duration]]=htmltools::plotTag(
  {ggsurv},
  sprintf("%s_KM",duration), width = 700, height = 400, pixelratio=3)
}
if(metaHeaders$best_response_perc!="NA"){
  my_comparisons=combn(levels(clean_RNA_sample_info$Cluster),2)%>% as.data.frame %>% as.list
  sink(NULL)
  p=ggpubr::ggboxplot(clean_RNA_sample_info, x = "Cluster",
                        y = "best_response_perc",
                        combine = TRUE,
                        # merge = "flip",
                        ylab = "Best response (%)",
                        add = "jitter",                               # Add jittered points
                        add.params = list(size = 0.5, jitter = 0.1),
                        color = "Cluster",
                        palette = "Dark2")+
  stat_compare_means(comparisons = my_comparisons)+ # Add pairwise comparisons p-value
  stat_compare_means(label.y = max(clean_RNA_sample_info$best_response_perc,na.rm = T)+diff(range(clean_RNA_sample_info$best_response_perc,na.rm = T))/3.5)+
  labs(title = "Best overall response")
  sink()
  plots[["best response (%)"]]=ggplotly(p)
}


tabs <- lapply(names(plots), function(name) {
shiny::tabPanel(name, plots[[name]])
})

do.call(shiny::tabsetPanel, tabs)
rm(list = setdiff(ls(),c("params","clean_RNA_sample_info","log2_expressions","metaHeaders")))
```

### Differential gene expression
Samples are spliced into two groups according to threshold (OS and/or PFS and/or Response). If quantitative response value is selected instead of categorical response value for response, 0 is set as splicing threshold.

```{r DGE-delta}
require(fgsea)
require(ggrepel)
cat.phenotype=setNames(c(params$OS_thresh,params$PFS_thresh,params$response_thresh),c("OS","PFS","Response"))
tmp=c(metaHeaders$OS,metaHeaders$PFS,metaHeaders$best_response)
cat.phenotype=cat.phenotype[!is.na(tmp)& tmp!="NA"]

#read oncokb
oncoKB=read.csv(system.file("extdata/cancerGeneList_oncoKB_lastupdate05192023.tsv",package="expr"),sep = "\t",header = T,check.names = F)
rownames(oncoKB)=oncoKB$`Hugo Symbol`
oncoKB=oncoKB%>%mutate(
    `Is Oncogene`=replace(`Is Oncogene`,`Is Oncogene`=="Yes",T),
    `Is Oncogene`=replace(`Is Oncogene`,`Is Oncogene`=="No",F),
    `Is Oncogene`=as.logical(`Is Oncogene`),
    `Is Tumor Suppressor Gene`=replace(`Is Tumor Suppressor Gene`,`Is Tumor Suppressor Gene`=="Yes",T),
    `Is Tumor Suppressor Gene`=replace(`Is Tumor Suppressor Gene`,`Is Tumor Suppressor Gene`=="No",F),
    `Is Tumor Suppressor Gene`=as.logical(`Is Tumor Suppressor Gene`),
    `OncoKB Annotated`=replace(`OncoKB Annotated`,`OncoKB Annotated`=="Yes",T),
    `OncoKB Annotated`=replace(`OncoKB Annotated`,`OncoKB Annotated`=="No",F),
    `OncoKB Annotated`=as.logical(`OncoKB Annotated`)
    )
#read cancer gene census
cancerGeneCensus=read.csv(system.file("extdata/concensus_Census_allWed_Jan_18_20_33_56_2023.tsv",package="expr"),sep = "\t",header = T,check.names = F)
rownames(cancerGeneCensus)=cancerGeneCensus$`Gene Symbol`
cancerGeneCensus$Hallmark=cancerGeneCensus$Hallmark %>% replace(.,.=="Yes",T) %>% replace(.,.=="",F) %>% as.logical()

#load pathways
load(system.file("extdata/pathway.nano_hallmarkv2022.1.RData",package = "expr"))
load(system.file("extdata/protein_coding_ensemble2symbol.RData",package = "expr"))
paramStr=comps_limma=DGE=DGE.plots=DGE.DT=volcanoPlots=gseaRess=split_groups=gseaRes.sig=gseaPlots=list()
for (phenotype in names(cat.phenotype)){
  
  # set group factor
  if (phenotype %in% c("OS","PFS")){
     if (cat.phenotype[phenotype]=="NA" | is.na(cat.phenotype[phenotype])){
       thresh=median(clean_RNA_sample_info[,phenotype],na.rm = T)
     }else{
       thresh=as.numeric(cat.phenotype[phenotype])
     }
    group=data.frame(as.factor(ifelse(clean_RNA_sample_info[,phenotype]>=thresh,"Long","Short")))
    dimnames(group)=list(rownames(clean_RNA_sample_info),phenotype)
    contrast="Long-Short"
    paramStr[[phenotype]]=sprintf("Spliting threshold: %s ",format(thresh,nsmall=3))
    split_groups[[phenotype]]=setNames(group[[phenotype]],rownames(group))
    }
  if (phenotype %in% c("Response")){
    if(cat.phenotype[phenotype]=="stable disease as good response"){
      group=data.frame(as.factor(clean_RNA_sample_info$best_response))
      dimnames(group)=list(rownames(clean_RNA_sample_info),phenotype)
      levels(group[[phenotype]])[grep("Partial|PR$|Stable|SD",levels(group[[phenotype]]),ignore.case = T)]="Good"
      levels(group[[phenotype]])[grep("Progressive|PD",levels(group[[phenotype]]),ignore.case = T)]="Bad"
      paramStr[[phenotype]]=sprintf("Consider %s.",cat.phenotype[phenotype])
    }else if(cat.phenotype[phenotype]=="stable disease as bad response"){
      group=data.frame(as.factor(clean_RNA_sample_info$best_response))
      dimnames(group)=list(rownames(clean_RNA_sample_info),phenotype)
      levels(group[[phenotype]])[grep("Partial|PR$",levels(group[[phenotype]]),ignore.case = T)]="Good"
      levels(group[[phenotype]])[grep("Progressive|PD$|Stable|SD$",levels(group[[phenotype]]),ignore.case = T)]="Bad"
      paramStr[[phenotype]]=sprintf("Consider %s.",cat.phenotype[phenotype])
    }else if(cat.phenotype[phenotype]=="NA"|is.na(cat.phenotype[phenotype])){
      group=data.frame(as.factor(ifelse(clean_RNA_sample_info$best_response_perc>=0,"Good","Bad")))
      dimnames(group)=list(rownames(clean_RNA_sample_info),phenotype)
      paramStr[[phenotype]]=sprintf("Spliting threshold: %s ",format(0,nsmall=3))
    }
    contrast="Good-Bad"
    split_groups[[phenotype]]=setNames(group[[phenotype]],rownames(group))
  }
  
  # DGE
  group=data.frame(Sample_ID=clean_RNA_sample_info$Sample_ID,group)
  group=group[complete.cases(group),]
  sink(NULL)
  comps_limma[[phenotype]]=comp_limma=dge_limma(
    expressions = log2_expressions[,group$Sample_ID], is_rawcount = F,is_logged = T,normalize = F,
    sample_frequency_threshold=0.5,
    clinic_info = group,
    ID_col = "Sample_ID",
    group_col = phenotype,
    contrasts = c(contrast),
    method ="limma_trend")
  sink()
  # format stats table
  stats=comp_limma$statistics %>% arrange(get(sprintf("%s:P.Value",contrast)))
  stats=cbind(stats,protein_coding_ensemble2symbol[match(rownames(stats),protein_coding_ensemble2symbol$Gene_Symbol),-1])
  
  tmp=apply(oncoKB[stats$Gene_Symbol,c("OncoKB Annotated","Is Oncogene","Is Tumor Suppressor Gene")], 1, function(x) {
    x[1]=!x[1]
    paste(c("not annotated","oncogene","TSG")[x],collapse = "; ")})
  tmp=tmp %>% replace(.,.=="NA; NA; NA",NA) %>% replace(.,.=="","other")
  stats=cbind(stats,oncoKB=tmp)
  tmp=apply(cancerGeneCensus[stats$Gene_Symbol,c("Tier","Hallmark")], 1, function(x)
    paste(paste("Tier",x[1],sep=""),"Hallmark"[x[2]]))
  tmp=tmp %>% replace(.,.=="TierNA NA",NA) %>% trimws()
  stats=cbind(stats,cancerGeneCensus=tmp)
  
  DGE[[phenotype]]=stats
  sigRow=stats[[sprintf("%s:P.Value",contrast)]]<0.05 & abs(stats[[sprintf("%s:logFC",contrast)]])>params$DGE_logFC_thresh
  DGE.DT[[phenotype]]= stats[sigRow,] %>%
    DT::datatable(
      filter = "top",
      options = list(
      # scrollX=T,scrollY=T,
        initComplete = JS("function(settings, json) {","$(this.api().table().header()).css({'font-size': '12px'});","}")),
      caption = htmltools::tags$caption(
        style = 'caption-side: top; text-align: left; color:black; font-size:150% ;',
                  sprintf("DGE on %s: %s",phenotype,contrast)) ) %>%
    DT::formatRound(columns = names(which(sapply(.$x$data,is.numeric))), digits = 2)%>%
    formatStyle(columns = colnames(.$x$data), `font-size` = '12px')
  # box plots
  if(sum(sigRow)>0){
    DGE.plots[[phenotype]]=as_widget(ggplotly(
      plotBoxPlot(
        data_mx = log2_expressions,sampleAttr = group,
        MOI = (stats[sigRow,]%>% rownames())[1:min(12,sum(sigRow))],
        header4x = phenotype,header4ID = "Sample_ID",header4color = phenotype) + 
        theme(legend.position = "none") +
        labs(title = "Top differentially expressed genes")
      ))
  }
  
  # generate volcano plots
  dff=stats %>% 
    mutate(diffexpressed=ifelse(
      get(sprintf("%s:P.Value",contrast))<0.05,
      ifelse(
        get(sprintf("%s:logFC",contrast))>params$DGE_logFC_thresh,"Up",
        ifelse(
          get(sprintf("%s:logFC",contrast))<(-params$DGE_logFC_thresh),
          "Down",
          "No")),
      "No")) %>%
    mutate(`-log10(P)`=-log10(get(sprintf("%s:P.Value",contrast)))) %>%
    mutate(label=Gene_Symbol)
  dff$label[!(dff$`-log10(P)`>(-log10(0.05))&dff$diffexpressed!="No")]=NA

  volcanoPlots[[phenotype]]=ggplot(
    data=dff,
    aes(x=.data[[sprintf("%s:logFC",contrast)]],
        y=`-log10(P)`,
        col=diffexpressed, label=label)) +
    geom_point() + 
    theme_minimal() +
    geom_text_repel() +
    scale_color_manual(name="expression",values = c("dodgerblue3", "gray50", "firebrick3")) +
    geom_vline(xintercept=c(-params$DGE_logFC_thresh, params$DGE_logFC_thresh), linetype=2) +
    geom_hline(yintercept=-log10(0.05), linetype=2) +
    labs(title = sprintf("Differentially expressed genes (p<0.05, |log2FC|> %s)",params$DGE_logFC_thresh))+
    xlab(sprintf("%s:logFC",contrast))
  
  # GSEA
    # generate gsea table
  rank=stats %>% #filter(p_value<0.05,abs(Difference_of_means)>paramss$meanDiff.thresh) %>%
    arrange(get(sprintf("%s:logFC",contrast))) %>% select(sprintf("%s:logFC",contrast)) %>%
    tibble::rownames_to_column("x") %>% tibble::deframe()
  gseaRess[[phenotype]]=gseaRes=fgsea(
    pathways=pathway,
    stats=rank,
    minSize = min(sapply(pathway, length)),
    gseaParam = 1)
  gseaRes.sig[[phenotype]]=gseaRes %>% filter(pval<0.05) %>% dplyr::slice(order(pval)) %>% 
    DT::datatable(
      filter = "top",
      options = list(
        # scrollX=T,scrollY=T,
        initComplete = JS("function(settings, json) {","$(this.api().table().header()).css({'font-size': '12px'});","}")),
    rownames = FALSE,
    caption = htmltools::tags$caption(
      style = 'caption-side: top; text-align: left; color:black; font-size:150% ;',
      sprintf("GSEA on %s: %s", phenotype, contrast))) %>% 
    DT::formatRound(columns = names(which(sapply(.$x$data,is.numeric))), digits = 2)%>%
    formatStyle(columns = colnames(.$x$data), `font-size` = '12px')

  # generate gsea plots
  fgseaRes=gseaRes %>% filter(pval<0.05) %>% arrange(NES)
  gseaPlots[[phenotype]]=plotGseaTable(pathway[fgseaRes$pathway], rank, fgseaRes, 
              gseaParam=1)
  
  
}
save(gseaRess,pathway,DGE,split_groups,file = sprintf("gsea_%s.RData","delta"))

# tabs
img.gsea=lapply(gseaPlots, function(x) {
  htmltools::plotTag(
  {x},
  sprintf("%s GSEA: significant pathway", names(x)), width = 700, height = 600, pixelratio=3)
})

img.DGE=lapply(volcanoPlots, function(x) {
  htmltools::plotTag(
  {x},
  sprintf("%s DGE: volcano plot", names(x)), width = 600, height = 400, pixelratio=3)
})

tabs <- lapply(names(cat.phenotype), function(name) {
  shiny::tabPanel(name, htmltools::tagList(
    list(
    tags$p(paramStr[[name]]),
    img.DGE[[name]],
    tags$h5(sprintf("DGE on %s: %s",phenotype,contrast),style="color:black;"),
    DGE.DT[[name]],
    DGE.plots[[name]],
    tags$hr(style = "border-top: 1px solid #ccc;"),
    tags$h5(sprintf("GSEA on %s: %s", phenotype, contrast),style="color:black;"),
    img.gsea[[name]],
    gseaRes.sig[[name]]
    )))
})

tagList(do.call(shiny::tabsetPanel, tabs))

rm(list = setdiff(ls(),c("params","clean_RNA_sample_info","log2_expressions","metaHeaders")))
```

